File: tests/normalize.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { normalizePathname, splitSegments, parseQuery } from "../src/normalize.js";

test("strips query string", () => {
  assert.equal(normalizePathname("/articles?page=1"), "/articles");
});

test("strips url fragment", () => {
  assert.equal(normalizePathname("/articles#comments"), "/articles");
});

test("strips both query and fragment", () => {
  assert.equal(normalizePathname("/articles?page=1&perPage=10#comments"), "/articles");
});

test("returns path untouched when no query or fragment", () => {
  assert.equal(normalizePathname("/articles/javascript"), "/articles/javascript");
});

test("converts null/undefined to empty string", () => {
  assert.equal(normalizePathname(null), "");
  assert.equal(normalizePathname(undefined), "");
});

test("coerces non-string input to string", () => {
  assert.equal(normalizePathname(123), "123");
});

test("splitSegments ignores leading and trailing slashes", () => {
  assert.deepEqual(splitSegments("/a/b/c/"), ["a", "b", "c"]);
});

test("splitSegments returns empty array for root", () => {
  assert.deepEqual(splitSegments("/"), []);
});

test("parseQuery returns empty object without query string", () => {
  assert.deepEqual(parseQuery("/articles"), {});
});

test("parseQuery returns empty object for empty query", () => {
  assert.deepEqual(parseQuery("/articles?"), {});
});

test("parseQuery parses single key as string", () => {
  assert.deepEqual(parseQuery("/articles?page=1"), { page: "1" });
});

test("parseQuery parses multiple keys", () => {
  assert.deepEqual(parseQuery("/articles?page=1&perPage=10"), {
    page: "1",
    perPage: "10",
  });
});

test("parseQuery collects repeated keys into an array", () => {
  assert.deepEqual(parseQuery("/search?tag=a&tag=b"), { tag: ["a", "b"] });
});

test("parseQuery ignores the url fragment", () => {
  assert.deepEqual(parseQuery("/articles?page=1#top"), { page: "1" });
});

test("parseQuery decodes url-encoded values", () => {
  assert.deepEqual(parseQuery("/search?q=hello%20world"), { q: "hello world" });
});

File: go/normalize_test.go
package roution

import (
	"reflect"
	"testing"
)

func TestNormalize(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want string
	}{
		{
			name: "strips query string",
			in:   "/articles?page=1",
			want: "/articles",
		},
		{
			name: "strips url fragment",
			in:   "/articles#comments",
			want: "/articles",
		},
		{
			name: "strips both query and fragment",
			in:   "/articles?page=1&perPage=10#comments",
			want: "/articles",
		},
		{
			name: "returns path untouched",
			in:   "/articles/javascript",
			want: "/articles/javascript",
		},
		{
			name: "empty string",
			in:   "",
			want: "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Normalize(tt.in)

			if got != tt.want {
				t.Fatalf("Normalize(%q) = %q, want %q", tt.in, got, tt.want)
			}
		})
	}
}

func TestSplitSegments(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want []string
	}{
		{
			name: "ignores leading and trailing slashes",
			in:   "/a/b/c/",
			want: []string{"a", "b", "c"},
		},
		{
			name: "returns empty slice for root",
			in:   "/",
			want: []string{},
		},
		{
			name: "ignores duplicate slashes",
			in:   "//a///b//",
			want: []string{"a", "b"},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := SplitSegments(tt.in)

			if !reflect.DeepEqual(got, tt.want) {
				t.Fatalf("SplitSegments(%q) = %#v, want %#v", tt.in, got, tt.want)
			}
		})
	}
}

func TestParseQuery(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want map[string]any
	}{
		{
			name: "returns empty map without query",
			in:   "/articles",
			want: map[string]any{},
		},
		{
			name: "returns empty map for empty query",
			in:   "/articles?",
			want: map[string]any{},
		},
		{
			name: "single key",
			in:   "/articles?page=1",
			want: map[string]any{
				"page": "1",
			},
		},
		{
			name: "multiple keys",
			in:   "/articles?page=1&perPage=10",
			want: map[string]any{
				"page":    "1",
				"perPage": "10",
			},
		},
		{
			name: "repeated keys",
			in:   "/search?tag=a&tag=b",
			want: map[string]any{
				"tag": []string{"a", "b"},
			},
		},
		{
			name: "ignores fragment",
			in:   "/articles?page=1#top",
			want: map[string]any{
				"page": "1",
			},
		},
		{
			name: "decodes url encoding",
			in:   "/search?q=hello%20world",
			want: map[string]any{
				"q": "hello world",
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ParseQuery(tt.in)

			if !reflect.DeepEqual(got, tt.want) {
				t.Fatalf("ParseQuery(%q) = %#v, want %#v", tt.in, got, tt.want)
			}
		})
	}
}